### DISCUSSION TITLE
UR ROS2 driver - Humble Release

### DISCUSSION THREAD

**Discussion by LukGen** [UniversalRobots/Universal_Robots_ROS2_Driver#431] [Q&A]
Hello,

Is there a timeline, when the UR-ROS2 driver will be released for humble?

**Comment by RobertWilbrandt** [comment]
The package has already been released for humble. It will be available as a binary package as soon as the next humble sync happens, which should be within the next few weeks.

**Reply by LukGen** [reply]
Will the released packages be branched out?
The humble packages don't build at the moment.

**Reply by RobertWilbrandt** [reply]
You can get the released version by checking out the 2.2.1 tag. Humble development will continue on the main branch together with rolling until there is a specific reason that this is no longer possible (e.g. a breaking change in rclcpp).

The current CI build failures are related to #412 and should be fixable by simply going back to the released version of ```ros2_controllers``` (which was the plan anyway). I will fix this shortly, but this does not affect released packages.

**Reply by LukGen** [reply]
Thank you for your reply. I now checked out the 2.2.1 Tag and build the released humble packages. However building the joint state broadcaster fails with the following message:

--- stderr: joint_state_broadcaster                                                                                           
In file included from /home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:15:
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:74:25: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
   74 |   controller_interface::CallbackReturn on_init() override;
      |                         ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:77:25: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
   77 |   controller_interface::CallbackReturn on_configure(
      |                         ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:81:25: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
   81 |   controller_interface::CallbackReturn on_activate(
      |                         ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:85:25: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
   85 |   controller_interface::CallbackReturn on_deactivate(
      |                         ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:49:23: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
   49 | controller_interface::CallbackReturn JointStateBroadcaster::on_init()
      |                       ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:103:23: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
  103 | controller_interface::CallbackReturn JointStateBroadcaster::on_configure(
      |                       ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:180:23: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
  180 | controller_interface::CallbackReturn JointStateBroadcaster::on_activate(
      |                       ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:206:23: error: ‘CallbackReturn’ in namespace ‘controller_interface’ does not name a type
  206 | controller_interface::CallbackReturn JointStateBroadcaster::on_deactivate(
      |                       ^~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp: In member function ‘bool joint_state_broadcaster::JointStateBroadcaster::init_joint_data()’:
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:240:42: error: ‘const class hardware_interface::LoanedStateInterface’ has no member named ‘get_prefix_name’
  240 |     if (name_if_value_mapping_.count(si->get_prefix_name()) == 0)
      |                                          ^~~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:242:34: error: ‘const class hardware_interface::LoanedStateInterface’ has no member named ‘get_prefix_name’
  242 |       name_if_value_mapping_[si->get_prefix_name()] = {};
      |                                  ^~~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:250:32: error: ‘const class hardware_interface::LoanedStateInterface’ has no member named ‘get_prefix_name’
  250 |     name_if_value_mapping_[si->get_prefix_name()][interface_name] = kUninitializedValue;
      |                                ^~~~~~~~~~~~~~~
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp: In member function ‘virtual controller_interface::return_type joint_state_broadcaster::JointStateBroadcaster::update(const rclcpp::Time&, const rclcpp::Duration&)’:
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:348:44: error: ‘const class hardware_interface::LoanedStateInterface’ has no member named ‘get_prefix_name’
  348 |     name_if_value_mapping_[state_interface.get_prefix_name()][interface_name] =
      |                                            ^~~~~~~~~~~~~~~
In file included from /opt/ros/humble/include/class_loader/class_loader/class_loader_core.hpp:57,
                 from /opt/ros/humble/include/class_loader/class_loader/class_loader.hpp:55,
                 from /opt/ros/humble/include/pluginlib/pluginlib/class_list_macros.hpp:40,
                 from /home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:400:
/opt/ros/humble/include/class_loader/class_loader/meta_object.hpp: In instantiation of ‘B* class_loader::impl::MetaObject<C, B>::create() const [with C = joint_state_broadcaster::JointStateBroadcaster; B = controller_interface::ControllerInterface]’:
/opt/ros/humble/include/class_loader/class_loader/meta_object.hpp:216:7:   required from here
/opt/ros/humble/include/class_loader/class_loader/meta_object.hpp:218:12: error: invalid new-expression of abstract class type ‘joint_state_broadcaster::JointStateBroadcaster’
  218 |     return new C;
      |            ^~~~~
In file included from /home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:15:
/home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:57:7: note:   because the following virtual functions are pure within ‘joint_state_broadcaster::JointStateBroadcaster’:
   57 | class JointStateBroadcaster : public controller_interface::ControllerInterface
      |       ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:24,
                 from /home/developer/ros_ur_driver/src/ros2_controllers/joint_state_broadcaster/src/joint_state_broadcaster.cpp:15:
/opt/ros/humble/include/controller_interface/controller_interface.hpp:93:50: note:     ‘virtual rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn controller_interface::ControllerInterface::on_init()’
   93 |   virtual LifecycleNodeInterface::CallbackReturn on_init() = 0;
      |                                                  ^~~~~~~
gmake[2]: *** [CMakeFiles/joint_state_broadcaster.dir/build.make:76: CMakeFiles/joint_state_broadcaster.dir/src/joint_state_broadcaster.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:178: CMakeFiles/joint_state_broadcaster.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake: *** [Makefile:146: all] Error 2
---
Failed   <<< joint_state_broadcaster [29.2s, exited with code 2]


Can you help me with that? Thank you in Advance
